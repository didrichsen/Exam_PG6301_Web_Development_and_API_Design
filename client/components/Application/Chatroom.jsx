import { useParams } from "react-router-dom";
import { useContext, useEffect, useMemo, useState } from "react";
import {
  ApiContext,
  fetchChatRoomById,
  updateChatroom,
} from "../../context/ApiContext";
import { configureWebSocket } from "../../utils/webSocket";
import HandleError from "../ErrorHandling/HandleError";

const Chatroom = () => {
  const { id } = useParams();
  const [chatroom, setChatroom] = useState(null);
  const [newComment, setNewComment] = useState("");
  const { user, fetchChatrooms, addComment } = useContext(ApiContext);
  const [webSocket, setWebSocket] = useState(null);
  const [isEditable, setIsEditable] = useState(false);
  const [newChatroomTitle, setNewChatroomTitle] = useState("");
  const [newDescription, setNewDescription] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const [loading, setLoading] = useState(true);

  const isValidTitle = useMemo(() => {
    return newChatroomTitle.length === 0;
  }, [newChatroomTitle]);

  const isValidMessage = useMemo(() => {
    return newComment.length === 0;
  }, [newComment]);

  const handleSubmitComment = async (e) => {
    e.preventDefault();

    const currentTime = new Date();
    const formattedDate = currentTime.toLocaleDateString("nb-NO");
    const formattedTime = currentTime.toLocaleTimeString("nb-NO", {
      hour: "numeric",
      minute: "numeric",
    });

    const craftedMessage = `"${newComment} " - ${user.name}(${user.email}), ${formattedDate} ${formattedTime}`;

    chatroom.comments.push(craftedMessage);

    const result = await addComment(craftedMessage, user, chatroom);

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      webSocket.send(
        JSON.stringify({
          type: "comment",
          review: craftedMessage,
        }),
      );

      setNewComment("");
    }
  };

  const handleNewComment = async (event) => {
    const data = JSON.parse(event.data);
    if (data.type === "comment") {
      await loadChatroom();
    }
  };

  const loadChatroom = async () => {
    const result = await fetchChatRoomById(id);

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      const chatroom = result.data;
      setChatroom(chatroom);
      setLoading(false);
    }
  };

  useEffect(() => {
    loadChatroom();
  }, []);

  useEffect(() => {
    const socket = configureWebSocket(handleNewComment);
    setWebSocket(socket);
  }, []);

  const handleEditToggle = () => {
    setIsEditable(!isEditable);
  };

  const handleUpdates = async () => {
    const result = await updateChatroom(newChatroomTitle, newDescription, id);

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      setNewChatroomTitle("");
      setNewDescription("");
      setIsEditable(false);
      await loadChatroom();
    }
  };

  if (loading) {
    return <div>loading ...</div>;
  }

  if (errorMessage) {
    return <HandleError errorMessage={errorMessage} />;
  }

  return (
    <>
      <div className="details-container">
        <div className="chatroom-details-container">
          <h2>
            {isEditable ? (
              <>
                <label htmlFor="newChatroomTitle">New Chatroom Title:</label>
                <input
                  type="text"
                  id={"newChatroomTitle"}
                  value={newChatroomTitle}
                  onChange={(e) => setNewChatroomTitle(e.target.value)}
                />
              </>
            ) : (
              chatroom.chatroomTitle
            )}
          </h2>
          <p>
            <strong>Admin</strong>: {chatroom.admin}
          </p>
          <p>
            {isEditable ? (
              <>
                <label htmlFor="newDescription">New Description: </label>
                <input
                  type="text"
                  id="newDescription"
                  value={newDescription}
                  onChange={(e) => setNewDescription(e.target.value)}
                />
              </>
            ) : (
              chatroom.description
            )}
          </p>
          {isEditable && (
            <button
              className="edit-and-save-button"
              onClick={handleUpdates}
              disabled={isValidTitle}
            >
              Save Changes
            </button>
          )}
          {user.email === chatroom.admin && (
            <button className="edit-and-save-button" onClick={handleEditToggle}>
              {isEditable ? "Cancel" : "Edit"}
            </button>
          )}

          <form onSubmit={handleSubmitComment}>
            <label>
              Write your Message:
              <input
                type={"text"}
                value={newComment}
                name="newComment"
                onChange={(e) => setNewComment(e.target.value)}
              />
            </label>
            <button disabled={isValidMessage}>Submit Message</button>
          </form>
        </div>

        <div className="message-container">
          {chatroom.comments.length === 0 && (
            <p>No comments yet in this chatroom. Be the first!</p>
          )}
          {chatroom.comments.map((comment, index) => (
            <div key={index} className="review">
              <p>{comment.comment}</p>
            </div>
          ))}
        </div>
      </div>
      )
    </>
  );
};

export default Chatroom;
