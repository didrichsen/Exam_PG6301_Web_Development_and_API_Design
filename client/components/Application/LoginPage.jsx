import LoginWithMicrosoft from "../Authentication/OpenIdConnect/ActiveDirectory/LoginWithMicrosoft";
import LoginWithGoogle from "../Authentication/OpenIdConnect/Google/LoginWithGoogle";

const LoginPage = ({ applicationConfig }) => {
  return (
    <div className="login-container">
      <button className="login-button">
        <LoginWithMicrosoft applicationConfig={applicationConfig}>
          Login With Microsoft
        </LoginWithMicrosoft>
      </button>
      <button className="login-button">
        <LoginWithGoogle applicationConfig={applicationConfig}>
          Login With Google
        </LoginWithGoogle>
      </button>
    </div>
  );
};

export default LoginPage;
