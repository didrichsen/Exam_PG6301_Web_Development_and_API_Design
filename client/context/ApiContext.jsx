import React from "react";

export const ApiContext = React.createContext({
  user: undefined,
  fetchChatrooms: async () => [],
  fetchUser: async () => {},
  createChatroom: async () => {},
  addComment: async () => {},
  fetchAllUsers: async () => {},
  loadUser: async () => {},
});

export async function fetchChatrooms() {
  try {
    const response = await fetch("/api/chatroom");

    if (!response.ok) {
      throw new Error("Something went wrong while fetching chatrooms.");
    }

    const data = await response.json();

    return { data, error: null };
  } catch (error) {
    return { error };
  }
}

export async function fetchUser() {
  try {
    const response = await fetch("/api/auth/login");

    if (!response.ok) {
      throw new Error("Something went wrong while fetching user.");
    }

    const data = await response.json();

    return { data, error: null };
  } catch (error) {
    return { error };
  }
}

export async function createChatroom(chatroom) {
  try {
    const response = await fetch("/api/chatroom/create", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ chatroom }),
    });

    if (!response.ok) {
      throw new Error("Something went wrong while posting chatroom.");
    }

    return { error: null };
  } catch (error) {
    return { error };
  }
}

export async function addComment(craftedMessage, user, chatroom) {
  try {
    const result = await fetch(`/api/chatroom/${chatroom._id}/comment`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ craftedMessage, user }),
    });

    if (!result.ok) {
      throw new Error("Something went wrong went posting comment.");
    }

    return { error: null };
  } catch (error) {
    return { error };
  }
}

export async function fetchAllUsers() {
  try {
    const response = await fetch("/api/user");

    if (!response.ok) {
      throw new Error("Something went wrong while fetching users.");
    }

    const data = await response.json();

    return { data, error: null };
  } catch (error) {
    return { error };
  }
}

export async function fetchChatRoomById(id) {
  try {
    const response = await fetch(`/api/chatroom/${id}`);

    if (!response.ok) {
      throw new Error("Something went wrong while fetching chatroom.");
    }

    const data = await response.json();

    return { data, error: null };
  } catch (error) {
    return { error };
  }
}

export async function updateUser(newName, newBio, user) {
  try {
    const result = await fetch(`/api/user/${user._id}`, {
      method: "PUT",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ newName, newBio }),
    });

    if (!result.ok) {
      throw new Error("Something went wrong when updating user");
    }

    return { error: null };
  } catch (error) {
    return { error };
  }
}

export async function updateChatroom(newChatroomTitle, newDescription, id) {
  try {
    const result = await fetch(`/api/chatroom/${id}`, {
      method: "PUT",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ newChatroomTitle, newDescription }),
    });

    if (!result.ok) {
      throw new Error("Something went wrong when updating chatroom");
    }

    return { error: null };
  } catch (error) {
    return { error };
  }
}
