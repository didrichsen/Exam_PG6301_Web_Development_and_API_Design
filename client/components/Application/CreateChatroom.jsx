import { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { ApiContext } from "../../context/ApiContext";
import { configureWebSocket } from "../../utils/webSocket";
import HandleError from "../ErrorHandling/HandleError";

const CreateChatroom = () => {
  const [chatroomTitle, setChatroomTitle] = useState("");
  const [description, setDescription] = useState("");
  const [webSocket, setWebSocket] = useState(null);
  const [isPrivate, setIsPrivate] = useState(false);
  const [isChatroomTitleTaken, setIsChatroomTitleTaken] = useState(false);
  const [participants, setParticipants] = useState([]);
  const [existingChatrooms, setExistingChatrooms] = useState([]);
  const [errorMessage, setErrorMessage] = useState("");

  const { createChatroom, user, fetchAllUsers, fetchChatrooms } =
    useContext(ApiContext);
  //User shall be added to selectedParticipants by default
  const [selectedParticipants, setSelectedParticipants] = useState([user]);

  //user.email is filtered out. We don't want to invite ourselves to the chatroom.
  const loadUsersAndChatrooms = async () => {
    let result = await fetchAllUsers();

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      const users = result.data;
      const usersToChooseFrom = users.filter((u) => u.email !== user.email);
      setParticipants(usersToChooseFrom);
    }

    //A list of chatroom titles is used to check if the chatroom title is taken.
    result = await fetchChatrooms();

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      const chatrooms = result.data;
      const chatroomTitles = chatrooms.map(
        (chatroom) => chatroom.chatroomTitle,
      );
      setExistingChatrooms(chatroomTitles);
    }
  };

  useEffect(() => {
    loadUsersAndChatrooms();
  }, []);

  const navigate = useNavigate();

  const handleSubmit = async (e) => {
    e.preventDefault();

    const chatRoom = {
      chatroomTitle: chatroomTitle,
      description: description,
      admin: user.email,
      participants: selectedParticipants,
      isPrivate: isPrivate,
    };

    const result = await createChatroom(chatRoom);

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      webSocket.send(
        JSON.stringify({
          type: "chatroomCreated",
          chatroomTitle: chatroomTitle,
          description: description,
        }),
      );
      setChatroomTitle("");
      setDescription("");
      setIsPrivate(false);
      navigate("/chatrooms");
    }
  };

  //Move participant from participants to selectedParticipants
  const handleParticipantChange = (participantId) => {
    const selectedParticipant = participants.find(
      (participant) => participant._id === participantId,
    );

    setSelectedParticipants((prevSelected) => [
      ...prevSelected,
      selectedParticipant,
    ]);

    setParticipants((prevParticipants) =>
      prevParticipants.filter(
        (participant) => participant._id !== participantId,
      ),
    );
  };

  //Checking to see if the chatroom title is taken
  const handleChatroomTitleChange = (value) => {
    const trimmedTittle = value.trim();
    setChatroomTitle(value);
    if (existingChatrooms.includes(trimmedTittle)) {
      setIsChatroomTitleTaken(true);
    } else {
      setIsChatroomTitleTaken(false);
    }
  };

  //Move participant from selectedParticipants to participants
  const handleParticipantRemoval = (participantId) => {
    const removedParticipant = selectedParticipants.find(
      (participant) => participant._id === participantId,
    );

    setParticipants((prevParticipants) => [
      ...prevParticipants,
      removedParticipant,
    ]);

    setSelectedParticipants((prevSelected) =>
      prevSelected.filter((participant) => participant._id !== participantId),
    );
  };

  //Configure websocket to be used when chatroom is created
  useEffect(() => {
    const socket = configureWebSocket();
    setWebSocket(socket);
  }, []);

  //Priting error message to user if something occur when creating chatroom.
  if (errorMessage) {
    return <HandleError errorMessage={errorMessage} />;
  }

  return (
    <div className="center-content-container">
      <div style={{ marginBottom: "2em" }}>
        <h2 style={{ textAlign: "center", marginBottom: "0.5em" }}>
          Let's get you a room!
        </h2>
        <h3>
          Create a room. Either <span style={{ color: "red" }}>private</span> or{" "}
          <span style={{ color: "green" }}>public</span>!
        </h3>
      </div>
      <form className="add-chatroom-form" onSubmit={handleSubmit}>
        <label>
          Enter Chatroom Title:
          <input
            type={"text"}
            value={chatroomTitle}
            name="title"
            onChange={(e) => handleChatroomTitleChange(e.target.value)}
          />
        </label>
        <label>
          Enter Chatroom Description:
          <textarea
            value={description}
            name="chatroomDescription"
            onChange={(e) => setDescription(e.target.value)}
            rows={5}
            style={{ width: "50em", marginTop: "2em" }}
          />
        </label>
        <label>
          Private Chatroom:
          <input
            type={"checkbox"}
            checked={isPrivate}
            name="isPrivate"
            onChange={(e) => setIsPrivate(e.target.checked)}
          />
        </label>
        {isPrivate ? (
          <>
            <label>
              Select Participants:
              <select multiple={true} name="participants">
                {participants.map((participant) => (
                  <option
                    key={participant._id}
                    value={participant._id}
                    onClick={() => handleParticipantChange(participant._id)}
                  >
                    {participant.name} ({participant.email})
                  </option>
                ))}
              </select>
            </label>
            <p>Participants selected:</p>
            {selectedParticipants.map((participant) => (
              <div key={participant._id}>
                {participant.name} ({participant.email})
                {participant.email === user.email ? (
                  "(You)"
                ) : (
                  <button
                    onClick={() => handleParticipantRemoval(participant._id)}
                  >
                    Remove
                  </button>
                )}
              </div>
            ))}
          </>
        ) : null}
        <button disabled={isChatroomTitleTaken}>Submit</button>
      </form>
    </div>
  );
};

export default CreateChatroom;
