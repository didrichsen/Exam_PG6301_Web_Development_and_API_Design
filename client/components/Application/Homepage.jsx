import { ApiContext } from "../../context/ApiContext";
import { useContext } from "react";

const Homepage = () => {
  const { user } = useContext(ApiContext);

  return (
    <div className="center-content-container">
      <h1>
        Welcome, {user.name} ({user.email})!
      </h1>
      <div className="container-welcome-message">
        <p>
          Explore and create chatrooms to connect with others who share your
          interests. Share thoughts, ideas, and engage in discussions on
          subjects that matter to you. Start by creating or joining a room and
          dive into meaningful conversations! You can create a chatroom by
          clicking on the "Create Chatroom" button in the navigation bar or you
          can browse through the list of chatrooms by clicking on the
          "Chatrooms" button.
        </p>
        <br></br>
        <p>
          Explore additional profiles to discover like-minded individuals who
          share your interests. Should you find a profile that resonates with
          your passions, take the next step by creating a chatroom and extending
          an invitation. Who knows, you might uncover more than just a shared
          interest â€” perhaps the beginning of a meaningful connection!
        </p>
      </div>
    </div>
  );
};

export default Homepage;
