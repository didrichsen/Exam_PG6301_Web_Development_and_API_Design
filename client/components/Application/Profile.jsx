import { useContext, useMemo, useState } from "react";
import { ApiContext, updateUser } from "../../context/ApiContext";
import HandleError from "../ErrorHandling/HandleError";

const Profile = () => {
  const [isEditable, setIsEditable] = useState(false);
  const [newName, setNewName] = useState("");
  const [newBio, setNewBio] = useState("");
  const [errorMessage, setErrorMessage] = useState("");

  const { user, loadUser } = useContext(ApiContext);

  const isValidCredentials = useMemo(() => {
    return newName.length === 0;
  }, [newName]);

  const handleEditToggle = () => {
    setIsEditable(!isEditable);
  };

  const handleSaveChanges = async () => {
    const result = await updateUser(newName, newBio, user);

    if (result.error) {
      setErrorMessage(result.error.message);
    } else {
      await loadUser();
      setIsEditable(false);
    }
  };

  if (errorMessage) {
    return <HandleError errorMessage={errorMessage} />;
  }

  return (
    <div className="center-content-container">
      <div>
        <strong>Your profile</strong>
      </div>
      <div>
        Username:{" "}
        {isEditable ? (
          <input
            type="text"
            value={newName}
            onChange={(e) => setNewName(e.target.value)}
          />
        ) : (
          user.name
        )}
      </div>
      <div>Email: {user.email}</div>
      <div>
        Bio:{" "}
        {isEditable ? (
          <input
            type="text"
            value={newBio}
            onChange={(e) => setNewBio(e.target.value)}
          />
        ) : (
          user.bio
        )}
      </div>
      {isEditable && (
        <button onClick={handleSaveChanges} disabled={isValidCredentials}>
          Save Changes
        </button>
      )}
      <button style={{ width: "10em" }} onClick={handleEditToggle}>
        {isEditable ? "Cancel" : "Edit"}
      </button>
    </div>
  );
};

export default Profile;
