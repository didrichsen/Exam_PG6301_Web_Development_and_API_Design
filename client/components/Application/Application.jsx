import { Route, Routes } from "react-router-dom";
import Navbar from "./NavBar";
import Logout from "../Authentication/Application/Logout";
import { useEffect, useState } from "react";
import Profile from "./Profile";
import ListChatrooms from "./ListChatrooms";
import CreateChatroom from "./CreateChatroom";
import Chatroom from "./Chatroom";
import {
  fetchChatrooms,
  fetchUser,
  createChatroom,
  addComment,
  fetchAllUsers,
  deleteChatroom,
    fetchChatRoomById,
    updateChatroom,
} from "../../context/ApiContext";
import { ApiContext } from "../../context/ApiContext";
import Homepage from "./Homepage";
import UserProfile from "./UserProfile";
import HandleError from "../ErrorHandling/HandleError";

const Application = () => {
  const [user, setUser] = useState({});
  const [errorMessage, setErrorMessage] = useState("");

  async function loadUser() {
    const response = await fetchUser();

    if (response.error) {
      setErrorMessage(response.error.message);
    } else {
      const data = response.data;
      setUser(data);
    }
  }

  useEffect(() => {
    loadUser();
  }, []);

  if (!user.name) {
    return <div className="center-content-container">Loading...</div>;
  }

  if (errorMessage) {
    return <HandleError errorMessage={errorMessage} />;
  }

  return (
    <>
      <Navbar />
      <ApiContext.Provider
        value={{
          fetchChatrooms,
          addComment,
          createChatroom,
          fetchUser,
          fetchAllUsers,
          loadUser,
          user,
          deleteChatroom,
            fetchChatRoomById,
            updateChatroom,
        }}
      >
        <main>
          <Routes>
            <Route path={"/"} element={<Homepage />} />
            <Route path={"/chatroom/:id"} element={<Chatroom />} />
            <Route path={"/users"} element={<UserProfile />} />

            <Route path={"/chatrooms"} element={<ListChatrooms />} />
            <Route path={"/chatroom-create"} element={<CreateChatroom />} />

            <Route path={"/profile"} element={<Profile />} />
            <Route path={"/logout"} element={<Logout />} />
            <Route path={"*"} element={<h2>Not Found</h2>} />
          </Routes>
        </main>
      </ApiContext.Provider>
      <footer>
        <div style={{ textAlign: "center", color: "white" }}>
          <p>Â© 2023 Exam Web- and Api Design</p>
          <p>Version 1.0.0</p>
        </div>
      </footer>
    </>
  );
};

export default Application;
