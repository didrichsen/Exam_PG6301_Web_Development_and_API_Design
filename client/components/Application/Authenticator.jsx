import { useEffect, useState } from "react";
import LoginCallback from "../Authentication/OpenIdConnect/Callback/LoginCallback";
import LoginPage from "./LoginPage";
import Application from "./Application";
import { BrowserRouter, Link } from "react-router-dom";

const Authenticator = () => {
  async function loadConfig() {
    const response = await fetch("/api/auth/config");
    if (!response.ok) {
      throw new Error(`Failed to fetch config:${response.status}`);
    }

    const config = await response.json();
    setApplicationConfig(config);
  }

  async function onLogin() {
    await loadConfig();
    window.location = "/";
  }

  const [applicationConfig, setApplicationConfig] = useState();

  useEffect(() => {
    loadConfig();
  }, []);

  if (!applicationConfig) {
    return <div>Loading...</div>;
  }

  if (window.location.pathname === "/login/callback") {
    return (
      <LoginCallback applicationConfig={applicationConfig} onLogin={onLogin} />
    );
  }

  if (!applicationConfig.user) {
    return (
      <>
        <div className="nav-container">
          <nav className="nav-bar">
            <div style={{ textAlign: "center", color: "white" }}>
              Welcome, Log in to create rooms and chat!
            </div>
          </nav>
        </div>
        <LoginPage applicationConfig={applicationConfig} />
        <footer>
          <div style={{ textAlign: "center", color: "white" }}>
            <p>Â© 2023 Exam Web- and Api Design</p>
            <p>Version 1.0.0</p>
          </div>
        </footer>
      </>
    );
  }

  return (
    <BrowserRouter>
      <Application />
    </BrowserRouter>
  );
};

export default Authenticator;
