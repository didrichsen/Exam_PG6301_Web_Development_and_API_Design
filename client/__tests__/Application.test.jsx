import renderer, { act } from "react-test-renderer";
import Application from "../components/Application/Application";
import { MemoryRouter } from "react-router-dom";
import { ApiContext } from "../context/ApiContext";

// Import fetchUser explicitly
import { fetchUser } from "../context/ApiContext";

jest.mock("../utils/webSocket");

jest.mock("../context/ApiContext", () => ({
  ...jest.requireActual("../context/ApiContext"),
  fetchUser: jest.fn(() =>
    Promise.resolve({
      data: { name: undefined },
      error: null,
    }),
  ),
}));

describe("Test of application component", () => {
  it('should show "loading..." if user is undefined', async () => {
    let component;

    await act(async () => {
      component = renderer.create(
        <MemoryRouter>
          <ApiContext.Provider value={{ fetchUser }}>
            <Application />
          </ApiContext.Provider>
        </MemoryRouter>,
      );
    });

    expect(component).toMatchSnapshot();
  });
});
