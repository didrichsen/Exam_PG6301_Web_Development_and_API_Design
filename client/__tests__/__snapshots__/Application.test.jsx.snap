// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of application component should show "loading..." if user is undefined 1`] = `
<div>
  Loading...
</div>
`;
